inference-engine/src/vpu/common/src/ngraph/operations/dynamic_shape_resolver.cpp

```diff
@@ -30,18 +34,159 @@ void DynamicShapeResolver::validate_and_infer_types() {
+bool getShapeFromHostTensorData(const HostTensorPtr& data, Shape& result) {
+    using T = typename element_type_traits<ET>::value_type;
+    T *dataPtr = data->get_data_ptr<ET>();
+    if (!dataPtr) {
+        return false;
+    }
+    size_t outputRank = data->get_shape()[0];
+
+    for (int i = 0; i < outputRank; i++) {
+        result.push_back(dataPtr[i]);
+    }
+
+    return true;
+}
+
+bool getShapeFromHostTensorData(const HostTensorPtr& data, Shape& shape) {
+    switch (data->get_element_type()) {
+        case element::Type_t::i8:
+            return getShapeFromHostTensorData<element::Type_t::i8>(data, shape);
+        case element::Type_t::i16:
+            return getShapeFromHostTensorData<element::Type_t::i16>(data, shape);
+        case element::Type_t::i32:
+            return getShapeFromHostTensorData<element::Type_t::i32>(data, shape);
+        case element::Type_t::i64:
+            return getShapeFromHostTensorData<element::Type_t::i64>(data, shape);
+        case element::Type_t::u8:
+            return getShapeFromHostTensorData<element::Type_t::u8>(data, shape);
+        case element::Type_t::u16:
+            return getShapeFromHostTensorData<element::Type_t::u16>(data, shape);
+        case element::Type_t::u32:
+            return getShapeFromHostTensorData<element::Type_t::u32>(data, shape);
+        case element::Type_t::u64:
+            return getShapeFromHostTensorData<element::Type_t::u64>(data, shape);
+        default:
+            return false;
+    }
+    return true;
+}
+
+template<element::Type_t InType>
+bool evaluate(const HostTensorPtr& inputTensor,
+              const HostTensorPtr& outShapeTensor,
+              const HostTensorPtr& outputTensor) {
+    Shape inputShape = inputTensor->get_shape();
+    Shape outputShape;
+    if (!getShapeFromHostTensorData(outShapeTensor, outputShape)) {
+        return false;
+    }
+
+    if (!ngraph::PartialShape(outputShape).refines(outputTensor->get_partial_shape())) {
+        return false;
+    }
+
+    outputTensor->set_shape(outputShape);
+
+    using T = typename element_type_traits<InType>::value_type;
+    T *inputPtr = inputTensor->get_data_ptr<InType>();
+    T *outputPtr = outputTensor->get_data_ptr<InType>();
+
+    const auto inTotalDimSize = shape_size(inputShape);
+    const auto strides = row_major_strides(inputShape);
+
+    size_t outputTensorIdx = 0;
+
+    for (size_t inputTensorIdx = 0; inputTensorIdx < inTotalDimSize; ++inputTensorIdx) {
+        auto offset = inputTensorIdx;
+        bool needCopy = true;
+        for (size_t dim = 0; dim < strides.size(); ++dim) {
+            const auto coordAlongDim = offset / strides[dim];
+            if (coordAlongDim > outputShape[dim] - 1) {
+                needCopy = false;
+                break;
+            }
+
+            offset %= strides[dim];
+        }
+        if (needCopy) {
+            outputPtr[outputTensorIdx++] = inputPtr[inputTensorIdx];
+        }
+    }
```

Comment:
Is it an algorithm to travers ND tensor element by element? If so, please extract it to a separate function - it's needed in multiple places

Reply:
Not really. This algorithm is applicable only in a special case when we are sure that the input data has larger each dimension than output and real shape is provided